Add `ParquetMetadataWriter` allow ad-hoc encoding of `ParquetMetadata`

adriangb opened the pull request:
A step towards #5988, https://github.com/apache/arrow-rs/issues/6002

alamb commented on parquet/src/file/footer.rs:
```diff
@@ -256,4 +313,68 @@ mod tests {
+
+        let encoded = encode_metadata(&metadata).unwrap();
+        let decoded = decode_metadata(&encoded).unwrap();
+        assert_eq!(
```
Can you simply just assert that `encoded` == `decoded`?

alamb commented on parquet/src/file/footer.rs:
```diff
@@ -256,4 +313,68 @@ mod tests {
+        {
+            assert_eq!(a, b);
+        }
+        // TODO: add encoding and decoding of column and offset indexes (aka page indexes)
```
I agree that encoding/decoding of these structures doesn't have to be present in the initial PR, however given they are stored out of line / slightly differently than the other structures I think it would be good to ensure we could encode them using this same API

alamb commented on parquet/src/file/footer.rs:
```diff
@@ -112,6 +109,38 @@ pub fn decode_metadata(buf: &[u8]) -> Result<ParquetMetaData> {
+/// specified by the [Parquet Spec].
+///
+/// [Parquet Spec]: https://github.com/apache/parquet-format#metadata
+pub fn encode_metadata(metadata: &ParquetMetaData) -> Result<Vec<u8>> {
```
Is it possible to switch the existing writers to use this API as well? Not only would that avoid code duplication, it would ensure the API is general enough

For example, I wonder if it would make sense for this function signature to be more like

```rust
/// write the metadata to the target `std::io:Write`, returning the number of bytes written
pub fn encode_metadata<W: Write>(metadata: &ParquetMetaData) -> Result<usize> {
...
}
```

That would allow writing into a `Vec` but also allow writing into various other targets and perhaps avoid buffering

alamb reviewed (commented):
Thanks @adriangb  -- this PR looks good to me and I think we could proceed with this design. 

I did file https://github.com/apache/arrow-rs/issues/6002 to track a potentially more flexible API that I think is worth considering. However, adding this API to mirror decode_metadata I think would also be fine (and we could make a more complex API later)

adriangb commented:
@alamb I pushed a fluentish API version of this.

I got bogged down implementing the page index writing because there doesn't seem to be a clean path to go from `ColumnIndex` and `OffsetIndex` to the thrift `Index` and [`ColumnIndex`](https://github.com/adriangb/arrow-rs/blob/afa975d13c824a158c500e721bdad8f2f9fc23f8/parquet/src/format.rs#L4483). I think the thing is that the current writers never materialize a `ParquetMetadata` and thus forcing them to do so might introduce unnecessary overhead. Maybe the path to go from a `ParquetMetadata` to bytes shouldn't be merged with writers? I do think the readers could be merged.

adriangb commented:
One thing I can do to avoid blocking on my lack of knowledge of encoding the page index stuff is to design the API first and implement it later. E.g. we can add `.with_page_index(bool)` and error if you set it to `true` or don't set it at all so that you're forced to acknowledge that the future default will be `true`.

alamb commented:
Thanks @adriangb  -- I will try and review this PR today

alamb commented:
Working through the list of PRs in arrow-rs is on my list of things to do tomorrow

alamb commented on parquet/src/file/metadata/mod.rs:
```diff
@@ -86,7 +86,7 @@ pub type ParquetOffsetIndex = Vec<Vec<Vec<PageLocation>>>;
 /// [`parquet.thrift`]: https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift
 /// [`parse_metadata`]: crate::file::footer::parse_metadata
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq)]
```
👍

alamb commented on parquet/src/file/footer.rs:
```diff
@@ -112,6 +109,46 @@ pub fn decode_metadata(buf: &[u8]) -> Result<ParquetMetaData> {
+        let column_orders = encode_column_orders(metadata.file_metadata().column_orders());
+        let schema = types::to_thrift(&metadata.file_metadata().schema().clone())?;
+
+        let t_file_metadata = TFileMetaData {
```
I noticed that this is not quite the same code as used n the actual writer (specifically the way column order is not the same) so I worry it would be inconsistent or drift over time from the actual writer

https://github.com/apache/arrow-rs/blob/22e0b4432c9838f2536284015271d3de9a165135/parquet/src/file/writer.rs#L352-L375

Thus what I suggest we do here is change writer.rs to use the `ParquetMetadataEncoder` and refactor the code from there into this function. That would be a bit more involved but I think would set us up nicely so that metadata encoding remains consistent.

alamb reviewed (commented):
Thanks @adriangb  -- this is looking like a good start

I think we should try and structure the code so the existing writer uses this new `MetadataEncoder` which would keep metadata writing consistent as well as enable usecases like encoding bloom filters, etc. 

Let me know what you think. 

cc @sunchao @tustvold @Jefffrey @liukun4515 @nevi-me   for any thoughts you might have on this API / approach

adriangb commented:
> I think we should try and structure the code so the existing writer uses this new MetadataEncoder which would keep metadata writing consistent as well as enable usecases like encoding bloom filters, etc.

I completely agree. That's just a much bigger chunk to bite off, I can give it a shot but I may need support to get there.

adriangb commented:
I've made some progress. I made a (very rough) metadata writer that is used internally by `SerializedFileWriter` and can encode from a `ParquetMetadata`. My plan of attack from here:
1. Implement reading of metadata without needing to have the entire file available. There's already `MetadataLoader` as pointed out in https://github.com/apache/arrow-rs/issues/6002#issuecomment-2208748759 but it wants to read metadata from an entire file and I think needs to be refactored to be able to load metadata when that's all you have.
2. Get feedback here on the APIs (they really aren't pretty).
3. Add roundtrip tests.

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -301,93 +277,36 @@ impl<W: Write + Send> SerializedFileWriter<W> {
+            Some(self.props.created_by().to_string()),
+            self.props.writer_version().as_num(),
+        );
+        encoder.finish()
```
👍

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -301,93 +277,36 @@ impl<W: Write + Send> SerializedFileWriter<W> {
             .map(|v| v.to_thrift())
             .collect::<Vec<_>>();
 
         self.write_bloom_filters(&mut row_groups)?;
```
FWIW https://github.com/apache/arrow-rs/pull/5933 also contains changes for bloom filter writing

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -791,23 +710,274 @@ impl<'a, W: Write + Send> PageWriter for SerializedPageWriter<'a, W> {
     }
 }
 
+struct ThriftMetadataWriter<'a, W: Write> {
```
I always get confused when reading the parquet code between what are the generated Thrift structures from the structures in https://docs.rs/parquet/latest/parquet/file/metadata/index.html

I like how you have split out writing of the thrift structures here from the writing of the `parquet::file` structures

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -791,23 +710,274 @@ impl<'a, W: Write + Send> PageWriter for SerializedPageWriter<'a, W> {
+        Ok(())
+    }
+
+    fn convert_column_indexes(&self) -> Vec<Vec<Option<ColumnIndex>>> {
```
I was looking around for another copy of this code and I now see that this is the first time we are going from `Index` --> `ColumnIndex` 

Makes sense to me. I think this type of structure could really help clean up some of the tests too (but I am getting ahead of myself)

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -301,93 +277,36 @@ impl<W: Write + Send> SerializedFileWriter<W> {
-
-        let file_metadata = parquet::FileMetaData {
-            num_rows,
+        let encoder = ThriftMetadataWriter::new(
```
This might read nicer like this:

```rust
let encoder = ThriftMetadataWriter::new()
  .with_schema(&self.schema)
  .with_descr(&self.descr)
  .with_row_groups(row_groups)
 ... 
  );

// encode the data to buf
encoder.encode(&mut buf)
```

Though I realize many of these fields are required

Maybe something like 

```rust
        let encoder = ThriftMetadataWriter::new(
            &self.schema,
            &self.descr,
             ...
        )
        .with_column_indexes(&self.column_indexes)
        .with_offset_indexes(&self.offset_indexes);
        encoder.encode(&mut buf)
```

alamb reviewed (commented):
I think this is looking quite nice @adriangb  and I think we should try and proceed with this approach.

I think it would be easier to make progress if we can work on the approach incrementally as multiple smaller PRs rather than one large one (it will be easier for me to give you timely feedback)

Also, it is probably good to know of https://github.com/apache/arrow-rs/pull/5486 from @etseidl which could conflict as we change the metadata.

Also https://github.com/apache/arrow-rs/pull/5933 from @progval 

Given we are now being careful about breaking changes (see https://github.com/apache/arrow-rs/blob/master/CONTRIBUTING.md#breaking-changes)  I am worried that these PRs will interact / cause conflicts with each other

What do you think of this idea: https://github.com/apache/arrow-rs/issues/6050 ?

etseidl commented on parquet/src/file/writer.rs:
```diff
@@ -791,23 +710,274 @@ impl<'a, W: Write + Send> PageWriter for SerializedPageWriter<'a, W> {
+        if let Some(row_group_offset_indexes) = self.metadata.offset_index() {
+            (0..self.metadata.row_groups().len())
+                .map(|rg_idx| {
+                    let column_indexes = &row_group_offset_indexes[rg_idx];
```
Minor nit: could this be named `offset_indexes`?

etseidl commented on parquet/src/file/page_index/index.rs:
```diff
@@ -168,6 +168,37 @@ impl<T: ParquetValueType> NativeIndex<T> {
+            .collect::<Option<Vec<_>>>()
+            .unwrap_or_else(|| vec![vec![]; self.indexes.len()]);
+
+        let null_counts = self
+            .indexes
+            .iter()
+            .map(|x| x.null_count())
+            .collect::<Option<Vec<_>>>()
+            .unwrap_or_else(|| vec![0; self.indexes.len()]);
```
While merging with #5486, I noticed this. IIUC, if on read the optional thrift `ColumnIndex::null_counts` is not present, then the `PageIndex::null_count` will be `None`. When converting back to a thrift `ColumnIndex`, it appears that this will convert the missing `null_counts` into a vector of `num_pages` zeros. I don't know if this is the correct behavior, mostly because the spec is (AFAICT) silent on the interpretation of a non-present `null_counts`. Is it not present as an optimization when there are no nulls, or is it not present due to a lack of information (say a V1 encoder doesn't keep null counts since the V1 page header doesn't require them). Due to that ambiguity I think `null_counts` here should be `None` if any or all of the `PageIndex::null_count` fields is `None`. Perhaps stop after the `collect()` and pass `null_counts` directly below.

alamb commented:
Update here is I plan to make a `53 dev` branch today so we can start getting this code merged and iterate on the API

alamb commented:
Hi @adriangb  -- I changed this PR to point at the `53.0.0-dev` branch. I plan to give it a careful review tomorrow and then I am thinking we can merge it and iterate over the course of a few PRs

Again, I am really sorry for the delay in reviewing. I think this is a really important feature but I have been overwhelmed with reviews for the last week or two

adriangb commented:
Thank you @alamb! No need to apologize; you have such a diverse and impactful contribution to open source, your time management is really quite inspiring. If anything I need to apologize for lagging on applying feedback. I will go over this PR and incorporate feedback (hopefully before your review tomorrow).

alamb reviewed (commented):
Here is how I suggest we proceed with this PR:

1. Let's create an example with the usecase described in https://github.com/apache/arrow-rs/issues/6002#issuecomment-2227009013 (I will try to do this later today). I think this will motivate how the API looks like
2. In parallel we could pull out some of the simple usability changes (like adding `PartialEq` and `pub use thrift` stuff into their own PR so we can merge that.

alamb commented:
I started on a basic example here: https://github.com/apache/arrow-rs/pull/6081 -- tomorrow I'll try and find time to try and rebase it on this PR and see if I can do what is needed

alamb commented on parquet/src/file/page_index/index.rs:
```diff
@@ -168,6 +168,38 @@ impl<T: ParquetValueType> NativeIndex<T> {
         })
     }
+
+    pub(crate) fn to_column_index(&self) -> ColumnIndex {
```
I think calling this method  `to_thrift` might be more consistent with other APIs like

https://docs.rs/parquet/latest/parquet/file/metadata/struct.RowGroupMetaData.html#method.to_thrift 

The naming is already pretty confusing

adriangb replied:
👍🏻

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -814,20 +734,289 @@ impl<'a, W: Write + Send> PageWriter for SerializedPageWriter<'a, W> {
     }
 }
 
+struct ThriftMetadataWriter<'a, W: Write> {
```
I was trying to articulate how this differs from the `to_thrift` functions on the structures

Perhaps its difference is something like

```suggestion
/// Writes `crate::file::metadata` structures to a thrift encdoded byte streams
///
/// This structure handles the details of writing the various parts of parquet
/// metadata into a byte stream. It is used to write the metadata into a 
/// parquet file and can also write metadata into other locations (such as a 
/// store of bytes).
///
/// This is somewhat trickey because the metadata is not store as a single inline
/// thrift struture. It can have several "out of band" structures such as the OffsetIndex
/// and BloomFilters which are stored separately whose locations are stored as offsets 
struct ThriftMetadataWriter<'a, W: Write> {
```

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -814,20 +734,289 @@ impl<'a, W: Write + Send> PageWriter for SerializedPageWriter<'a, W> {
+    buf: &'a mut TrackedWrite<W>,
+    schema: &'a TypePtr,
+    schema_descr: &'a SchemaDescPtr,
+    row_groups: Vec<RowGroup>,
```
Rather than storing these fields separately, I wonder if it would be possible simply to store a https://docs.rs/parquet/latest/parquet/file/metadata/struct.FileMetaData.html

(can be done as a follow on PR)

Something like
```rust
struct ThriftMetadataWriter<'a, W: Write> {
    buf: &'a mut TrackedWrite<W>,
    parquet_metadata: &ParquetMetadata, // or maybe Arc<ParquetMetadata> 🤔 
}
```

And then add the various builder APIs like `with_key_value_metadata` directly to `ParquetMetadata` 

This would make it easier to work / manipulate `ParquetMetadata` in general and would make the responsibility of the writer clearer (handle the details of coordinating the writing of the thrift encoded structures and their indexes)

adriangb replied:
I agree, but lets do that in a followup PR since (I think) that would not be breaking in any way and this is quite large already

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -814,20 +734,289 @@ impl<'a, W: Write + Send> PageWriter for SerializedPageWriter<'a, W> {
+        Ok(())
+    }
+
+    /// Assembles and writes metadata at the end of the file.
```
```suggestion
    /// Assembles and writes the final metadata to self.buf
```

alamb reviewed (commented):
Thanks again @adriangb -- I think this is looking really close

While looking through the API for https://github.com/apache/arrow-rs/issues/6097 I had a few more suggestions, but then I think this would be ready. What do you think @etseidl ? I feel this is close to what is proposed  in https://github.com/apache/arrow-rs/pull/6095 and related PRs. I feel we are quite close to having some sort of reasonable API for reading / writing these structures:

* Metadata is stored in `ParquetFileMetadata` and associated structures
* Write `ParquetFileMetadata` to bytes using `ThriftMetadataWriter` (and maybe there will be an async version)
* (Eventually) we can have an equivalent `ThriftMetadataReader`  (and the async version `[MetadataLoader](https://docs.rs/parquet/latest/parquet/arrow/async_reader/struct.MetadataLoader.html))

etseidl commented:
> What do you think @etseidl ? I feel this is close to what is proposed in #6095 and related PRs.

Yes, I think this is ready to merge into 53.0.0-dev. The unencoded size info is still separate from the page locations, so hopefully this will merge in fairly cleanly. From a practical standpoint, I think it will be easier to merge this before #6095, and then I can make the needed changes to the new offset index struct to implement `to_thrift()`.

adriangb commented:
I think I've addressed the feedback and updated the branch :)

etseidl commented on parquet/src/file/writer.rs:
```diff
@@ -1854,6 +2053,163 @@ mod tests {
+        let decoded_metadata = load_metadata_from_bytes(metadata.file_size, data).await;
+
+        assert_eq!(metadata.metadata, decoded_metadata);
+
```
```suggestion
    }

```
Lost a brace during merge.

adriangb replied:
Thank you!

etseidl commented on parquet/src/file/writer.rs:
```diff
@@ -815,20 +735,299 @@ impl<'a, W: Write + Send> PageWriter for SerializedPageWriter<'a, W> {
+                    let offset_indexes = &row_group_offset_indexes[rg_idx];
+                    offset_indexes
+                        .iter()
+                        .map(|column_index| Some(OffsetIndex::new(column_index.clone(), None)))
```
After merging with the latest 53.0.0-dev, this can then become:
```suggestion
                        .map(|offset_index| Some(offset_index.to_thrift()))
```

adriangb commented:
I'm not sure why the test is failing (it was before, I don't think it's from a merge). Need to investigate.

etseidl commented:
> I'm not sure why the test is failing (it was before, I don't think it's from a merge). Need to investigate.

I think you'll need to merge 53.0.0-dev again to pick up the latest changes to the offset index, and then reformat (some new names are longer and changed how the linter wants lines wrapped).

adriangb commented:
I've updated the branch and cleaned up, test is still failing. It seems the reading part is trying to access byte 0 of the file, which doesn't make sense and makes me think there's a bug somewhere (could be in the test since there's a lot of shim in there): https://github.com/apache/arrow-rs/actions/runs/10082832978/job/27878006690?pr=6000#step:6:761

etseidl commented on parquet/src/file/writer.rs:
```diff
@@ -1854,6 +2053,167 @@ mod tests {
+
+        let data = buf.into_inner().freeze();
+
+        let decoded_metadata = load_metadata_from_bytes(metadata.file_size, data).await;
```
```suggestion
        let decoded_metadata = load_metadata_from_bytes(data.len(), data).await;
```
This will load the page indexes, but then the assert below fails because the `offset_index_offset` and `column_index_offset` fields of the column chunk are different. Might have to write an `equals` that accounts for that.

etseidl replied:
> I could write a custom eq but... that is going to be a pain.

True.  Perhaps just compare parts of the metadata rather than the whole thing. Let me see if I can whip something up quick and submit a PR to your branch.

> I also need to think about the implications of these things not matching up. If I understand correctly those are the offsets to the page index data from the start of the file (or from where?) and because we loaded the metadata from only a portion of the file they got re-calculated differently?

Well, you've written the page indexes to a new file (well, buffer), so those offsets point to their location in the new file as opposed to the old.  Looking at the test output, the column index is at offset 0 with length 23, and the offset index is at 23 with length 10.  Assuming there's no 'PAR1' header on the new file, those offsets seem correct.

alamb commented:
I merged the 53 dev branch and that seems to have closed this PR -- any chance you can reopen and retarget main?

alamb commented:
I wrote up some thoughts that were floating in my head in https://github.com/apache/arrow-rs/issues/6129 

I am hoping to spend some more time today looking at this PR in deatil

Thank you again for your patience

alamb commented:
I know I keep leading you along with comments on this PR. I really believe this is an important PR / API to work on, but at the moment i don't have enough bandwidth to help drive this forward / break it down into smaller pieces / make sure they all fit together. Maybe @etseidl  can figure out how to make it happen.

In any event I hope to have more time to devote here by the end of August

etseidl commented:
@adriangb I submitted a PR to fix the failing parquet and clippy tests. 

The "check compilation" test failure is beyond me. The "--no-default-features" flag causes some dependencies to not be loaded. Getting that one fixed requires some cargo mojo that I lack :(

etseidl commented:
@adriangb just wanted to warn you not to try merging with master just yet. A PR that merged to 53.0.0-dev after 53.0.0-dev was merged to master (but before 53.0.0-dev was merged to this branch) is present here. If you merge with master now that PR will be lost. Hopefully this will be corrected soon 😅.

adriangb commented:
@etseidl  looks like this was closed because the target branch was deleted. My first thought is that I should retarget this at master/main and rebase it on that but I guess that's what you're warning me not to do above? Not sure what the next steps are here then.

etseidl commented:
> My first thought is that I should retarget this at master/main and rebase it on that but I guess that's what you're warning me not to do above? Not sure what the next steps are here then.

If you retarget master, then there will be merge conflicts around the new histogram stuff I added. Those are pretty easy to clear (you can look at what I did in https://github.com/adriangb/arrow-rs/pull/3). Then I think this will be ready for a final review! Thanks!
